File: apps/backend-py/app/case_source.py
"""Runtime case-source selection for real Phoenix vs local Docker sandbox cases."""

from __future__ import annotations

from typing import Literal
from urllib.parse import urlparse

from .config import settings

CaseSourceSelection = Literal["real_erp", "sandbox_cases"]
ErpSource = Literal["real_erp", "local_or_mock", "sandbox_cases"]

_LOCAL_HOSTS = {
    "localhost",
    "127.0.0.1",
    "0.0.0.0",
    "::1",
    "host.docker.internal",
    "phoenix-mock",
}
_selected: CaseSourceSelection = (
    "sandbox_cases" if settings.sandbox_case_count > 0 else "real_erp"
)


def sandbox_available() -> bool:
    return settings.sandbox_case_count > 0


def selected_case_source() -> CaseSourceSelection:
    return _selected


def set_case_source(source: str) -> CaseSourceSelection:
    global _selected
    if source not in ("real_erp", "sandbox_cases"):
        raise ValueError("case source must be real_erp or sandbox_cases")
    if source == "sandbox_cases" and not sandbox_available():
        raise ValueError(
            "sandbox cases are not seeded; set SANDBOX_CASE_COUNT > 0 and restart compose"
        )
    _selected = source  # type: ignore[assignment]
    return _selected


def active_phoenix_base_url() -> str:
    if _selected == "sandbox_cases":
        return settings.sandbox_phoenix_api_base_url
    return settings.phoenix_api_base_url


def erp_source() -> ErpSource:
    if _selected == "sandbox_cases":
        return "sandbox_cases"
    host = (urlparse(settings.phoenix_api_base_url).hostname or "").lower()
    return "local_or_mock" if host in _LOCAL_HOSTS else "real_erp"


def status() -> dict:
    return {
        "case_source": _selected,
        "erp_source": erp_source(),
        "sandbox_case_count": settings.sandbox_case_count,
        "sandbox_available": sandbox_available(),
    }

File: backend-node/src/caseSource.ts
import { config } from "./config";

export type CaseSourceSelection = "real_erp" | "sandbox_cases";
export type ErpSource = "real_erp" | "local_or_mock" | "sandbox_cases";

const localHosts = new Set(["localhost", "127.0.0.1", "0.0.0.0", "::1", "host.docker.internal", "phoenix-mock"]);
let selected: CaseSourceSelection = config.sandboxCaseCount > 0 ? "sandbox_cases" : "real_erp";

export function sandboxAvailable(): boolean {
  return config.sandboxCaseCount > 0;
}

export function selectedCaseSource(): CaseSourceSelection {
  return selected;
}

export function setCaseSource(source: string): CaseSourceSelection {
  if (source !== "real_erp" && source !== "sandbox_cases") {
    throw new Error("case source must be real_erp or sandbox_cases");
  }
  if (source === "sandbox_cases" && !sandboxAvailable()) {
    throw new Error("sandbox cases are not seeded; set SANDBOX_CASE_COUNT > 0 and restart compose");
  }
  selected = source;
  return selected;
}

export function activePhoenixBaseUrl(): string {
  return selected === "sandbox_cases" ? config.sandboxPhoenixBaseUrl : config.realPhoenixBaseUrl;
}

export function erpSource(): ErpSource {
  if (selected === "sandbox_cases") return "sandbox_cases";
  const host = new URL(config.realPhoenixBaseUrl || "http://localhost").hostname.toLowerCase();
  return localHosts.has(host) ? "local_or_mock" : "real_erp";
}

export function caseSourceStatus() {
  return {
    case_source: selected,
    erp_source: erpSource(),
    sandbox_case_count: config.sandboxCaseCount,
    sandbox_available: sandboxAvailable(),
  };
}
